README.md

```diff
@@ -7,6 +7,128 @@
 
 `package dig` provides an opinionated way of resolving object dependencies.
 
+## Status
+
+BETA. Expect potential API changes.
+
+## Conatiner
+
+`package dig` exposes `type Container` as an object capable of resolving a
+directional dependency graph.
+
+To create one:
+```go
+import "go.uber.org/dig"
+
+func main() {
+	c := dig.New()
+	// dig container `c` is ready to use!
+}
+```
+
+Objects in the container are identified by their `reflect.Type` and **everything
+is treated as a singleton**, meaning there can be only one object in the graph
+of a given type.
+
+There are plans to expand the API to support multiple objects of the same type
+in the container, but for time being consider using a factory pattern.
+
+## Provide
+
+The `Provide` method adds an object, or a constructor of an object, to the container.
+
+### Provide a constructor
+
+Constructor can be a function returning any number of objects and, optionally,
+an error.
+
+Each argument to the constructor is registered as a **dependency** in the graph.
+
+```go
+type A struct {}
+type B struct {}
+type C struct {}
+
+c := dig.New()
+constructor := func (*A, *B) *C {
+  // At this point, *A and *B have been resolved through the graph
+  // and can be used to provide an object of type *C
+  return &C{}
+}
+err := c.Provide(constructor)
+// dig container is now able to provide *C to any constructor.
+//
+// However, note that in the current example *A and *B
+// have not been provided, meaning that the resolution of type
+// *C will result in an error, because types *A and *B can not
+// be instantiated.
+```
+
+### Provide an object
+
+As a shortcut for any object without dependencies, register it directly.
+
+```go
+type A struct {
+	Name string
+}
+
+c := dig.New()
+err := c.Provide(&A{Name: "Hello, dig!"})
+```
+
+## Invoke
+
+The `Invoke` API is the flip side of `Provide` and used to retrieve types from the container.
+
+`Invoke` looks through the graph and resolves all the constructor parameters for execution.
+
+In order to successfully use use `Invoke`, the function must meet the following criteria:
+
+1. Input to the `Invoke` must be a function
+1. All arguments to the function must be types in the container
+1. If an error is returned from an invoked function, it will be propagated to the caller
+
+Here is a fully working somewhat real-world `Invoke` example:
+
+```go
+package main
+
+import (
+	"go.uber.org/config"
```

Comment:
I'd suggest for examples to be self contained, so it will be easier to test them.